pkg/controller/garbagecollector/metrics.go

```diff
@@ -65,9 +75,47 @@ func Register() {
 		prometheus.MustRegister(EventProcessingLatency)
 		prometheus.MustRegister(DirtyProcessingLatency)
 		prometheus.MustRegister(OrphanProcessingLatency)
+		prometheus.MustRegister(DeletionToWatchLatency)
 	})
 }
 
 func sinceInMicroseconds(clock clock.Clock, start time.Time) float64 {
 	return float64(clock.Since(start).Nanoseconds() / time.Microsecond.Nanoseconds())
 }
+
+func NewDeletionToWatchTracker(limit int) *uidToDeletionTime {
+	return &uidToDeletionTime{
+		items: make(map[types.UID]time.Time),
+		lock:  sync.Mutex{},
+		limit: limit,
+	}
+}
+
+type uidToDeletionTime struct {
+	items map[types.UID]time.Time
+	lock  sync.Mutex
+	limit int
+}
+
+// DeletionSent should be called when the deletion request is sent.
+func (u *uidToDeletionTime) DeletionSent(uid types.UID, start time.Time) error {
+	u.lock.Lock()
+	defer u.lock.Unlock()
+	if len(u.items) >= u.limit {
```

Comment:
Suppose that GC for some reason tries to delete the same object twice, such
that the response from the first request is being processed when it makes
the second deletion request-- any chance that can happen?

On Fri, Aug 12, 2016 at 1:59 PM, Chao Xu <notifications@github.com> wrote:

> In pkg/controller/garbagecollector/metrics.go
> <https://github.com/kubernetes/kubernetes/pull/30483#discussion_r74659348>
> :
>
> > +		lock:  sync.Mutex{},
> > +		limit: limit,
> > +	}
> > +}
> > +
> > +type uidToDeletionTime struct {
> > +	items map[types.UID]time.Time
> > +	lock  sync.Mutex
> > +	limit int
> > +}
> > +
> > +// DeletionSent should be called when the deletion request is sent.
> > +func (u *uidToDeletionTime) DeletionSent(uid types.UID, start time.Time) error {
> > +	u.lock.Lock()
> > +	defer u.lock.Unlock()
> > +	if len(u.items) >= u.limit {
>
> I expect that not to happen. If the deletion request succeeds and we never
> see it through watch, that's a serious problem.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/30483/files/b1a45b0f84709c3d600d99009797c6a2e1ef1791#r74659348>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngllCSSL9RWm7IZ9ESt1fAvI0c91Yhks5qfN6egaJpZM4Jioy1>
> .
>